### PR TITLE
Fix #2270 - PDB parsing fixes

### LINKED ISSUE #2270: Rizin errors/crashes when parsing a specific PDB generated by MSVC2019
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Windows x64 64
| File format of the file you reverse (mandatory)      | PE
| Architecture/bits of the file (mandatory)            | x64
| `rizin -v` full output, **not truncated** (mandatory)  | rizin 0.4.0-git @ windows-x86-64 commit: f2e421a896e8502724b36d427b9913ca900516c6, build: 29/01/2022__11:48:41.71

### Expected behavior
Running `idp` on the given binary and symbols file successfully loads the debug symbols.

### Actual behavior
Rizin errors/crashes with the error:
```
ERROR: parse_type_fieldlist: Unsupported leaf type 0x1404
ERROR: Parse TPI type error. idx in stream: 0x1f65
```

### Steps to reproduce the behavior
<details><summary><code>main.cpp</code></summary>

```cpp
#include <iostream>
#include <stdexcept>

#ifdef _MSC_VER
#include <Windows.h>
#endif

#ifdef _MSC_VER
#define NOINLINE __declspec(noinline)
#else
#define NOINLINE __attribute__((noinline))
#endif

static NOINLINE void trycatch_cpp() {
    std::string message("Test");

    try {
        throw std::runtime_error(message);
    } catch (const std::runtime_error& e) {
        std::cout  << __func__ << ": Caught: " << e.what() << std::endl;
    }
}

#ifdef _MSC_VER
static NOINLINE void trycatch_seh_access_violation() {
    int* x = nullptr;
    __try {
        std::cout << *x << std::endl;
    } __except (GetExceptionCode() == EXCEPTION_ACCESS_VIOLATION) {
        std::cout << __func__ << ": Caught EXCEPTION_ACCESS_VIOLATION" << std::endl;
    }
}

static NOINLINE void trycatch_seh_access_violation_finally() {
    int* x = nullptr;
    __try {
        __try {
            std::cout << *x << std::endl;
        } __except (GetExceptionCode() == EXCEPTION_ACCESS_VIOLATION) {
            std::cout << __func__ << ": Caught EXCEPTION_ACCESS_VIOLATION" << std::endl;
        }
    } __finally {
        std::cout << __func__ << ": Finally" << std::endl;
    }
}

static NOINLINE void trycatch_outputdebugstring() {
    constexpr const char* MESSAGE = "Hello, World!\n";

    ULONG_PTR args[2];
	args[0] = std::strlen(MESSAGE) + 1;
	args[1] = reinterpret_cast<ULONG_PTR>(MESSAGE);

	__try {
		RaiseException(DBG_PRINTEXCEPTION_C, 0, _countof(args), args);
	} __except (GetExceptionCode() == DBG_PRINTEXCEPTION_C ? EXCEPTION_EXECUTE_HANDLER : EXCEPTION_CONTINUE_SEARCH) {
		std::cout << __func__ << ": Caught" << std::endl;
	}
}
#endif

int main() {
    trycatch_cpp();

#ifdef _MSC_VER
    trycatch_seh_access_violation();
    trycatch_seh_access_violation_finally();
    trycatch_outputdebugstring();
#endif
}
```
</details>

<details><summary><code>CMakeLists.txt</code></summary>

```cmake
cmake_minimum_required(VERSION 3.22)
project(exceptions-test)
add_executable(exceptions-test main.cpp)
```
</details>

Binary (Debug, MSVC 2019 16.11.9): [exceptions-test.zip](https://github.com/rizinorg/rizin/files/7963732/exceptions-test.zip)

P.S. This is a small piece of code I written to see how rizin handles analyzing exception handlers, might be useful as a test binary for anyone who will try to improve rizin in this area.

### Additional Logs, screenshots, source code,  configuration dump, ...

screenshot

### PR DESCRIPTION
**Pull request by GustavoLCR**
# Please don't squash
 

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

* Fix parsing if there's an empty stream
* Fix incomplete definitions due to not parsing LF_INDEX leaf types
* Fix incomplete definitions due to not redefining existing types
* Fix NULL derefs if parsing fails when freeing the `RzPdb` struct
* Do not call `exit()` if parsing a type fails
* Other small fixes

**Test plan**

Tests pass

**Closing issues**

Fixes https://github.com/rizinorg/rizin/issues/2270

### CODE REVIEW

File: librz/analysis/type_pdb.c [hunk 1]
```diff
@@ -407,6 +407,40 @@ static RzTypeStructMember *parse_struct_member(const RzTypeDB *typedb, RzPdbTpiS
 	return NULL;
 }
 
+static inline const char *base_type_kind_as_string(RzBaseTypeKind kind) {
```
Comment (XVilka) [inline librz/analysis/type_pdb.c]: Please move this function and make an API to `librz/type/base.c`.

File: librz/bin/pdb/pdb.c [hunk 2]
```diff
@@ -349,14 +353,9 @@ RZ_API void rz_bin_pdb_free(RzPdb *pdb) {
...
-	RZ_FREE(pdb->s_dbi);
 	free_gdata_stream(pdb->s_gdata);
-	RZ_FREE(pdb->s_gdata);
 	free_omap_stream(pdb->s_omap);
-	RZ_FREE(pdb->s_omap);
 	free_tpi_stream(pdb->s_tpi);
-	RZ_FREE(pdb->s_tpi);
 	free_pe_stream(pdb->s_pe);
-	RZ_FREE(pdb->s_pe);
 	RZ_FREE(pdb);
```
Comment (XVilka) [inline librz/bin/pdb/pdb.c]: Can be just `free(pdb)` as well.

**Review by XVilka — CHANGES_REQUESTED** [review]
A typo:
```
../librz/analysis/type_pdb.c:467: error: ',' expected (got "": Redefining type %s.\n"")
```

**Comment by XVilka** [comment]
@GustavoLCR since you are working on PDB, could you please take a quick look at https://github.com/rizinorg/rizin/issues/2270?
It makes sense to bundle the fix here if it's not too big.

**Comment by ret2libc** [comment]
Hey @GustavoLCR could you provide a more detailed summary of the PR? What kind of fixes are these?

**Review by XVilka — CHANGES_REQUESTED** [review]
1. Please rebase on top of the latest `dev`.
2. Please fix clang-format:
```
Ubuntu clang-format version 13.0.1-++20220120110924+75e33f71c2da-1~exp1~20220120231001.58
librz/type/base.c:28:4: error: code should be clang-formatted [-Wclang-format-violations]
/**
   ^
librz/type/base.c:29:59: error: code should be clang-formatted [-Wclang-format-violations]
* \brief Returns string representing the kind of base type
                                                          ^
librz/type/base.c:30:2: error: code should be clang-formatted [-Wclang-format-violations]
* 
 ^
librz/type/base.c:31:64: error: code should be clang-formatted [-Wclang-format-violations]
* \param kind RzBaseTypeKind to return string representation of
                                                               ^
```
3. There are two broken tests:
```
[XX] db/tools/rz_bin rz-bin -Pj
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '!rz-bin -Pj ${RZ_FILE}~{}
?e
' bins/pdb/minimal.pdb
-- stdout
--- expected
+++ actual
@@ -26,7 +26,7 @@
         },
         {
           "type": "union",
-          "name": "<unnamed-tag>",
+          "name": "type_0x1004",
           "size": 4,
           "members": [
             {
@@ -47,7 +47,7 @@
           "size": 8,
           "members": [
             {
-              "member_type": "union <unnamed-tag>",
+              "member_type": "union type_0x1004",
               "member_name": "a",
               "offset": 0
             },

-- stderr
Warning: Cannot find base address, flags will probably be misplaced
Warning: Cannot find base address, flags will probably be misplaced

[XX] db/cmd/cmd_idp idpij
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'idpij bins/pdb/minimal.pdb~{}' =
-- stdout
--- expected
+++ actual
@@ -25,7 +25,7 @@
       },
       {
         "type": "union",
-        "name": "<unnamed-tag>",
+        "name": "type_0x1004",
         "size": 4,
         "members": [
           {
@@ -46,7 +46,7 @@
         "size": 8,
         "members": [
           {
-            "member_type": "union <unnamed-tag>",
+            "member_type": "union type_0x1004",
             "member_name": "a",
             "offset": 0
           },

-- stderr
Warning: Cannot find base address, flags will probably be misplaced
Warning: Cannot find base address, flags will probably be misplaced
```
4. Please fix the following warning:
```
../librz/analysis/type_pdb.c:432:23: error: implicit conversion from enumeration type 'RzBaseTypeKind' to different enumeration type 'RzTypeIdentifierKind' [-Werror,-Wenum-conversion]
        t->identifier.kind = kind;
                           ~ ^~~~
1 error generated.
```

**Review by XVilka — APPROVED** [review]

**Pull request merged**